pkg/xds/experimental/client/listener.go

```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package client
+
+import (
+	"log/slog"
+
+	"github.com/cilium/cilium/pkg/envoy/xds"
+)
+
+// listener is a helper structure for watchers focused on handling a single, registered callback.
+type listener struct {
+	ID      uint64
+	typeUrl string
+	cb      WatcherCallback
+	trigger chan struct{}
+
+	log       *slog.Logger
+	resources xds.ResourceSource
+}
+
+var _ xds.ResourceVersionObserver = (*listener)(nil)
+
+// HandleNewResourceVersion implements xds.ResourceVersionObserver.
+// It triggers the callback process.
+func (l *listener) HandleNewResourceVersion(typeUrl string, _ uint64) {
+	if typeUrl != l.typeUrl {
+		return
+	}
+	select {
+	case l.trigger <- struct{}{}:
```

Comment:
Can this result in a missed update?

We should put comments on what kinds of guarantees we are making between a wake up from l.tigger and resource.GetResources() that makes this correct.

Reply:
Added comment to explain that it can't (due to buffered channel)